src/dmd/glue.d

```diff
@@ -1163,8 +1163,67 @@ void FuncDeclaration_toObjFile(FuncDeclaration fd, bool multiobj)
 
     if (fd.isCrtCtorDtor & 1)
         objmod.setModuleCtorDtor(s, true);
+
     if (fd.isCrtCtorDtor & 2)
-        objmod.setModuleCtorDtor(s, false);
+    {
+        bool mitigation = true;
+        /*
+            https://issues.dlang.org/show_bug.cgi?id=22520
+
+            Apple radar: https://openradar.appspot.com/FB9733712
+
+            Apple deprecated the mechanism used to implement `crt_constructor`
+            on MacOS Monterey. This works around that by generating a new function
+            (crt_destructor_thunk_NNN, run as a constructor) which registers
+            the destructor-to-be using __cxa_atexit()
+
+            This workaround may need a further look at when it comes to
+            shared library support, however there is no bridge for
+            that spilt milk to flow under yet.
+
+            This relies on the Itanium ABI so is portable to any
+            platform it, if needed.
+        */
+        if(mitigation)
+        {
+            __gshared uint nthDestructor = 0;
+            char* buf = cast(char*) calloc(50, 1);
+            assert(buf);
+            const ret = snprintf(buf, 100, "_dmd_crt_destructor_thunk_%u", nthDestructor++);
+            assert(ret >= 0 && ret < 100, "snprintf either failed or overran buffer");
+            //Function symbol
+            auto newConstructor = symbol_calloc(buf);
+            //Build type
+            newConstructor.Stype = type_function(TYnfunc, [], false, type_alloc(TYvoid));
+            //Tell it it's supposed to be a C function. Does it do anything? Not sure.
+            newConstructor.Stype.Tmangle = mTYman_c;
+            symbol_func(newConstructor);
+            //Global SC for now.
+            newConstructor.Sclass = SCglobal;
+            func_t* funcState = newConstructor.Sfunc;
+            //Init start block
+            funcState.Fstartblock = block_calloc();
+            block* startBlk = funcState.Fstartblock;
+            //Make that block run __cxa_atexit(&func);
+            auto atexitSym = getRtlsym(RTLSYM.CXA_ATEXIT);
+            //Build parameter pack - __cxa_atexit(&func, null, null)
+            auto paramPack = el_params(el_long(TYnptr, 0), el_long(TYnptr, 0), el_ptr(s), null);
```

Comment:
To make it friendly for shared libraries:
```
__cxa_atexit(&func, null, &__dso_handle);
```
`__dso_handle` itself is a weak symbol defined in every at least every dso object that cares about Itanium/C++ compatibility.